W-12689952: add ibmmq binding in async21+

arielmirra opened the pull request:
- W-12689952: add ibmmq message binding in async21+
- W-12689952: add ibmmq server binding in async21+

arielmirra committed:
chore: add multiple key/value pairs adding to syntax

arielmirra committed:
chore: add async2x plugin to unique platform tests

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async21Syntax.scala:
```diff
@@ -1,8 +1,14 @@
+    add(
+      Async20Syntax.nodes,
+      "bindings" -> Set(Mercure),
+      Mercure    -> Set(),
```
The "closed shpae" validation is already being thrown in the `validateEmptyMap` method of the `EmptyBindingParser` class. you should decide between one mechanism or the other.

arielmirra replied:
done, removed the empty sets

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async21Syntax.scala:
```diff
@@ -1,8 +1,14 @@
+      Async20Syntax.nodes,
+      "bindings" -> Set(Mercure),
+      Mercure    -> Set(),
+      "ibmmq"    -> Set()
```
if this is a mercure bindings commit, add only mercure binding related things.

arielmirra replied:
done, reorganized commits

tomsfernandez commented on amf-cli/shared/src/test/resources/validations/async20/validations/mercure-binding-nonempty.yaml:
```diff
@@ -7,25 +7,21 @@ servers:
     protocol: mercure
     bindings:
-      mercure: {
+      mercure:
```
This commits should be squashed into the one that introduced this file as it is a fix over the same file with the same content.

arielmirra replied:
done, reorganized commits

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/ibmmq/IBMMQMessageBinding.scala:
```diff
@@ -0,0 +1,51 @@
+package amf.apicontract.client.platform.model.domain.bindings.ibmmq
+
+import amf.apicontract.client.platform.model.domain.bindings.{BindingVersion, MessageBinding}
```
add all ibmmq things in the same commit as it is confusing to have the support for IBMMQ in two separate commits.

arielmirra replied:
done, reorganized commits

arielmirra committed:
W-12689952: add mercure empty fields validation

arielmirra committed:
chore: fix imports in async server binding parser

arielmirra committed:
W-12689952: add ibmmq message binding in async21+

- add message binding model, parser and emitter
- add unit validation and cycle tests
- add model to APIEntities and model.md

arielmirra committed:
W-12689952: add ibmmq server binding in async21+

- add server binding model, parser and emitter
- add unit validation and cycle tests
- add model to APIEntities and model.md

arielmirra committed:
W-12689952: add ibmmq channel binding in async21+

- add channel binding model, parser and emitter
- add unit validation and cycle tests
- add model to APIEntities and model.md

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/IBMMQBindingModel.scala:
```diff
@@ -0,0 +1,62 @@
+  val Description: Field =
+    Field(
+      Str,
+      ApiBinding + "description",
```
I think we already have a description propertyTerm here. Maybe we can reuse this one. Maybe ask Pope?

arielmirra replied:
done, we're now using the core description field

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/IBMMQBindingModel.scala:
```diff
@@ -0,0 +1,62 @@
+      ModelDoc(ModelVocabularies.ApiBinding, "type", "The type of the message")
+    )
+
+  val Headers: Field =
```
headers is a comma separate list of strings, maybe we should parse it? This field could change to be of type Array(StringType)

arielmirra replied:
done

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiMessageBindingsEmitter.scala:
```diff
@@ -121,3 +124,26 @@ class Amqp091MessageEmitter(binding: Amqp091MessageBinding, ordering: SpecOrderi
+) extends AsyncApiCommonBindingEmitter {
+  override def emit(b: EntryBuilder): Unit = {
+    b.entry(
+      YNode("ibmmq"),
```
You could use the Constant here.

arielmirra replied:
done 👌🏽

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/message/IBMMQMessageBindingParser.scala:
```diff
@@ -0,0 +1,27 @@
+    val binding = IBMMQMessageBinding(Annotations(entry))
+    val map     = entry.value.as[YMap]
+
+    map.key("type", IBMMQMessageBindingModel.MessageType in binding)
```
Watch out because most of these fields have default values.

arielmirra replied:
done, now we're parsing the value or setting the default one where applicable

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/server/IBMMQServerBindingParser.scala:
```diff
@@ -0,0 +1,28 @@
+    val binding = IBMMQServerBinding(Annotations(entry))
+    val map     = entry.value.as[YMap]
+
+    map.key("groupId", IBMMQServerBindingModel.GroupId in binding)
```
Watch out because most of these facets have default values.

arielmirra replied:
done, now we're parsing the value or setting the default one where applicable

tomsfernandez commented on amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/channel/IBMMQChannelBindingParser.scala:
```diff
@@ -0,0 +1,74 @@
+
+object IBMMQChannelBindingParser extends BindingParser[IBMMQChannelBinding] {
+  override def parse(entry: YMapEntry, parent: String)(implicit ctx: AsyncWebApiContext): IBMMQChannelBinding = {
+    val binding = IBMMQChannelBinding(Annotations(entry))
```
Same as the rest of the parsers, most of these facet have default values.

arielmirra replied:
done, now we're parsing the value or setting the default one where applicable

tomsfernandez commented:
@arielmirra what about binding validations ?? Is this being tracked in another issue?

tomsfernandez commented:
@arielmirra  Something I forgot in the review, you should check with Ale if the queue and topic field values of the ibmmq channel binding will be open or closed.

arielmirra committed:
W-12689952: use constants in binding emission

arielmirra commented:
@tomsfernandez we've added an issue to add binding validations later on, as they require much more work.

arielmirra committed:
W-12689952: add default values to ibmmq binding parsers

arielmirra committed:
W-12689952: use core description field in ibmmq binding description

arielmirra committed:
W-12689952: parse ibmmq message binding headers as a list of strings
